plugin/pathvirt/test/pv.c

```diff
@@ -0,0 +1,308 @@
+ * happen.
+ *
+ * If didn't return NULL, the returned pointer must be freed.
+ */
+static char *dynamic_path_swap(const char *path)
+{
+    /* quickly return NULL if no swap */
+    if (!should_swap) {
+        return NULL;
+    }
+
+    /* yes, should swap */
+
+    /* check if path is in list of registered paths to swap out */
+    int index = clfind(old_path_prefix_list, path);
+    if (index == -1)
+        return NULL;
+
+    /* found it in old list, now get a pointer to the new prefix to swap in*/
+    char *new = clget(new_path_prefix_list, index);
+    if (new == NULL)
+        return NULL;
+
+    size_t new_element_sz = clgetsize_ptr(new_path_prefix_list, new);
+    size_t old_element_sz = clgetsize_ind(old_path_prefix_list, index);
+
+    /* temporarily null terminate new element */
+    new[new_element_sz] = '\0';
+
+    /* finally, create full path with the new prefix swapped in */
+
+    /* plus 1 is for safety slash we include between the new prefix and the
+       unchanged rest of the path. this is in case their environment
+       variable doesn't end with a slash. in the "worst" case,
+       there will be two extra slashes if the new prefix ends with a slash
+       and the old one doesn't. plus 1 for NULL */
+    size_t newpathsize = (strlen(path) - old_element_sz) + new_element_sz + 1 + 1;
+    char *newpath = malloc(newpathsize);
+    snprintf(newpath, newpathsize, "%s/%s", new, path + old_element_sz);
+
+    /* repair the colon list */
+    new[new_element_sz] = ':';
+
+    return newpath;
+}
+
+/*
+ * Libc Hooks (for all path related functions)
+ */
+
+int fopen64(const char *path, const char *mode)
+{
+    char *hook_path = dynamic_path_swap(path);
+
+    /* hook_path was NULL, not swapping */
+    if (!hook_path)
+        return NEXT_FNC(fopen64)(path, mode);
+
+    /* swapping */
+    int fd = NEXT_FNC(fopen64)(hook_path, mode);
+
+    /* dynamic_path_swap's return val needs to be free'd */
+    free(hook_path);
+
+    return fd;
+}
+
+int open(const char *path, int oflag, mode_t mode)
+{
+    char *hook_path = dynamic_path_swap(path);
+
+    /* hook_path was NULL, not swapping */
+    if (!hook_path)
+        return NEXT_FNC(open)(path, oflag, mode);
+
+    /* swapping */
+    int fd = NEXT_FNC(open)(hook_path, oflag, mode);
+
+    /* dynamic_path_swap's return val needs to be free'd */
+    free(hook_path);
+
+    return fd;
+}
+
+/*
+ * DMTCP Setup
+ */
+
+void dmtcp_event_hook(DmtcpEvent_t event, DmtcpEventData_t *data)
+{
+    /* NOTE:  See warning in plugin/README about calls to printf here. */
+    switch (event) {
+    case DMTCP_EVENT_INIT:
+    {
+        /* On init, check if they've specified paths to virtualize via
+           DMTCP_PATH_PREFIX env */
+        char *old_env = getenv(ENV_DPP);
+        if (old_env) {
+
+            prefix_list_sz = strlen(old_env) + 1;
+            old_path_prefix_list = malloc(prefix_list_sz);
+            new_path_prefix_list = malloc(prefix_list_sz);
```

Comment:
I'm a little worried about using `malloc`, since the end-user code could also be using `malloc`.  Some other possibilities are using jalib/jalloc.h, or else statically allocating 10 KB for each prefix_list, using `#define PREFIX_LIST_SIZE (10*1024)`, and then `JASSERT( prefix_list_sz >= PREFIX_LIST_SIZE).(prefix_list_sz).text("Too many environment variables.  Please increase PREFIX_LIST_SIZE")`